File: caps-sdk-python/caps_sdk/db/postgres.py
import psycopg2
from psycopg2 import pool
import logging

logger = logging.getLogger('caps')


class PostgresManager:
    def __init__(self, config: dict = None):
        self.config = config or {
            "host": os.environ.get("CAPS_PG_HOST", "localhost"),
            "port": int(os.environ.get("CAPS_PG_PORT", "5432")),
            "user": os.environ.get("CAPS_PG_USER", "caps"),
            "password": os.environ.get("CAPS_PG_PASSWORD", "caps"),
            "database": os.environ.get("CAPS_PG_DB", "caps_platform"),
        }
        import os as _os
        self._pool = None
        self.connected = False

    def connect(self):
        self._pool = pool.ThreadedConnectionPool(
            minconn=1, maxconn=self.config.get("pool_size", 10), **self.config
        )
        conn = self._pool.getconn()
        conn.close()
        self._pool.putconn(conn)
        self.connected = True
        logger.info("[caps] PostgreSQL connected")

    def execute(self, query: str, params: tuple = None) -> list:
        if not self._pool:
            raise RuntimeError("PostgreSQL not connected")
        conn = self._pool.getconn()
        try:
            with conn.cursor() as cur:
                cur.execute(query, params)
                if cur.description:
                    return cur.fetchall()
                conn.commit()
                return []
        finally:
            self._pool.putconn(conn)

    def disconnect(self):
        if self._pool:
            self._pool.closeall()
            self.connected = False
            logger.info("[caps] PostgreSQL disconnected")

File: sdk-node/src/db/postgres.ts
import { Pool, PoolClient, QueryResult } from 'pg';

export class PostgresManager {
  private pool: Pool | null = null;
  private connected = false;
  private retryCount = 0;
  private maxRetries = 5;
  private config: any;

  constructor(config: any = {}) {
    this.config = {
      host: config.host || process.env.CAPS_PG_HOST || 'localhost',
      port: config.port || parseInt(process.env.CAPS_PG_PORT || '5432', 10),
      user: config.user || process.env.CAPS_PG_USER || 'caps',
      password: config.password || process.env.CAPS_PG_PASSWORD || 'caps',
      database: config.database || process.env.CAPS_PG_DB || 'caps_platform',
      max: config.poolSize || 10,
      idleTimeoutMillis: 30000,
      connectionTimeoutMillis: 5000,
    };
  }

  async connect(): Promise<void> {
    try {
      this.pool = new Pool(this.config);
      await this.pool.query('SELECT 1');
      this.connected = true;
      this.retryCount = 0;
      console.log('[caps] PostgreSQL connected');
    } catch (err: any) {
      console.error('[caps] PostgreSQL connection failed:', err.message);
      this.connected = false;
      this.scheduleReconnect();
    }
  }

  async query(text: string, params?: any[]): Promise<QueryResult> {
    if (!this.pool) throw new Error('PostgreSQL not connected');
    return this.pool.query(text, params);
  }

  async getClient(): Promise<PoolClient> {
    if (!this.pool) throw new Error('PostgreSQL not connected');
    return this.pool.connect();
  }

  private scheduleReconnect(): void {
    if (this.retryCount >= this.maxRetries) return;
    const delay = Math.min(1000 * Math.pow(2, this.retryCount), 30000);
    this.retryCount++;
    setTimeout(() => this.connect(), delay);
  }

  async disconnect(): Promise<void> {
    if (this.pool) {
      await this.pool.end();
      this.connected = false;
      console.log('[caps] PostgreSQL disconnected');
    }
  }

  get isConnected(): boolean { return this.connected; }
  get health(): { connected: boolean; poolSize: number } {
    return {
      connected: this.connected,
      poolSize: this.config.max,
    };
  }
}
